File: stripe/src/test/java/com/stripe/android/view/ShippingInfoWidgetTest.java
package com.stripe.android.view;

import android.support.design.widget.TextInputLayout;
import android.view.View;

import com.stripe.android.BuildConfig;
import com.stripe.android.R;
import com.stripe.android.model.Address;
import com.stripe.android.model.ShippingInformation;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.MockitoAnnotations;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.android.controller.ActivityController;
import org.robolectric.annotation.Config;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertTrue;

/**
 * Test class for {@link ShippingInfoWidget}
 */
@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class, sdk = 25)
public class ShippingInfoWidgetTest {

    private ShippingInfoWidget mShippingInfoWidget;
    private TextInputLayout mAddressLine1TextInputLayout;
    private TextInputLayout mAddressLine2TextInputLayout;
    private TextInputLayout mCityTextInputLayout;
    private TextInputLayout mNameTextInputLayout;
    private TextInputLayout mPostalCodeTextInputLayout;
    private TextInputLayout mStateTextInputLayout;
    private StripeEditText mAddressLine1EditText;
    private StripeEditText mAddressLine2EditText;
    private StripeEditText mPostalEditText;
    private StripeEditText mCityEditText;
    private StripeEditText mNameEditText;
    private StripeEditText mStateEditText;
    private StripeEditText mPhoneEditText;
    private CountryAutoCompleteTextView mCountryAutoCompleteTextView;

    private String mNoPostalCodeCountry = "ZW"; // Zimbabwe
    private Address mAddress;
    private ShippingInformation mShippingInfo;

    @Before
    public void setup() {
        MockitoAnnotations.initMocks(this);
        Locale.setDefault(Locale.US);
        ActivityController<ShippingInfoTestActivity> activityController =
                Robolectric.buildActivity(ShippingInfoTestActivity.class).create().start();
        mShippingInfoWidget = activityController.get().getShippingInfoWidget();
        mAddressLine1TextInputLayout = mShippingInfoWidget.findViewById(R.id.tl_address_line1_aaw);
        mAddressLine2TextInputLayout = mShippingInfoWidget.findViewById(R.id.tl_address_line2_aaw);
        mCityTextInputLayout = mShippingInfoWidget.findViewById(R.id.tl_city_aaw);
        mNameTextInputLayout = mShippingInfoWidget.findViewById(R.id.tl_name_aaw);
        mPostalCodeTextInputLayout = mShippingInfoWidget.findViewById(R.id.tl_postal_code_aaw);
        mStateTextInputLayout = mShippingInfoWidget.findViewById(R.id.tl_state_aaw);
        mAddressLine1EditText = mShippingInfoWidget.findViewById(R.id.et_address_line_one_aaw);
        mAddressLine2EditText = mShippingInfoWidget.findViewById(R.id.et_address_line_two_aaw);
        mCityEditText = mShippingInfoWidget.findViewById(R.id.et_city_aaw);
        mNameEditText = mShippingInfoWidget.findViewById(R.id.et_name_aaw);
        mPostalEditText = mShippingInfoWidget.findViewById(R.id.et_postal_code_aaw);
        mStateEditText = mShippingInfoWidget.findViewById(R.id.et_state_aaw);
        mPhoneEditText = mShippingInfoWidget.findViewById(R.id.et_phone_number_aaw);
        mCountryAutoCompleteTextView = mShippingInfoWidget.findViewById(R.id.country_autocomplete_aaw);
        mAddress = new Address.Builder()
                .setCity("San Francisco")
                .setState("CA")
                .setCountry("US")
                .setLine1("185 Berry St")
                .setLine2("10th Floor")
                .setPostalCode("12345")
                .build();
        mShippingInfo = new ShippingInformation(mAddress, "Fake Name", "(123) 456 - 7890");
    }

    @Test
    public void shippingInfoWidget_whenCountryChanged_fieldsRenderCorrectly() {
        mCountryAutoCompleteTextView.updateUIForCountryEntered(Locale.US.getDisplayCountry());
        assertEquals(mAddressLine1TextInputLayout.getHint(), mShippingInfoWidget.getResources().getString(R.string.address_label_address));
        assertEquals(mAddressLine2TextInputLayout.getHint(), mShippingInfoWidget.getResources().getString(R.string.address_label_apt_optional));
        assertEquals(mPostalCodeTextInputLayout.getHint(), mShippingInfoWidget.getResources().getString(R.string.address_label_zip_code));
        assertEquals(mStateTextInputLayout.getHint(), mShippingInfoWidget.getResources().getString(R.string.address_label_state));

        mCountryAutoCompleteTextView.updateUIForCountryEntered(Locale.CANADA.getDisplayCountry());
        assertEquals(mAddressLine1TextInputLayout.getHint(), mShippingInfoWidget.getResources().getString(R.string.address_label_address));
        assertEquals(mAddressLine2TextInputLayout.getHint(), mShippingInfoWidget.getResources().getString(R.string.address_label_apt_optional));
        assertEquals(mPostalCodeTextInputLayout.getHint(), mShippingInfoWidget.getResources().getString(R.string.address_label_postal_code));
        assertEquals(mStateTextInputLayout.getHint(), mShippingInfoWidget.getResources().getString(R.string.address_label_province));

        mCountryAutoCompleteTextView.updateUIForCountryEntered(Locale.UK.getDisplayCountry());
        assertEquals(mAddressLine1TextInputLayout.getHint(), mShippingInfoWidget.getResources().getString(R.string.address_label_address_line1));
        assertEquals(mAddressLine2TextInputLayout.getHint(), mShippingInfoWidget.getResources().getString(R.string.address_label_address_line2_optional));
        assertEquals(mPostalCodeTextInputLayout.getHint(), mShippingInfoWidget.getResources().getString(R.string.address_label_postcode));
        assertEquals(mStateTextInputLayout.getHint(), mShippingInfoWidget.getResources().getString(R.string.address_label_county));

        mCountryAutoCompleteTextView.updateUIForCountryEntered(new Locale("", mNoPostalCodeCountry).getDisplayCountry());
        assertEquals(mAddressLine1TextInputLayout.getHint(), mShippingInfoWidget.getResources().getString(R.string.address_label_address_line1));
        assertEquals(mAddressLine2TextInputLayout.getHint(), mShippingInfoWidget.getResources().getString(R.string.address_label_address_line2_optional));
        assertEquals(mPostalCodeTextInputLayout.getVisibility(), View.GONE);
        assertEquals(mStateTextInputLayout.getHint(), mShippingInfoWidget.getResources().getString(R.string.address_label_region_generic));
    }

    @Test
    public void shippingInfoWidget_addressSaved_validationTriggers() {
        mCountryAutoCompleteTextView.updateUIForCountryEntered(Locale.US.getDisplayCountry());
        assertFalse(mShippingInfoWidget.validateAllFields());
        mAddressLine1EditText.setText("123 Fake Address");
        mNameEditText.setText("Fake Name");
        mCityEditText.setText("Fake City");
        mPostalEditText.setText("12345");
        mStateEditText.setText("CA");
        mPhoneEditText.setText("(123) 456 - 7890");
        assertTrue(mShippingInfoWidget.validateAllFields());
        mPostalEditText.setText("");
        assertFalse(mShippingInfoWidget.validateAllFields());
        mPostalEditText.setText("ABCDEF");
        assertFalse(mShippingInfoWidget.validateAllFields());
        mCountryAutoCompleteTextView.updateUIForCountryEntered(new Locale("", mNoPostalCodeCountry).getDisplayCountry());
        assertTrue(mShippingInfoWidget.validateAllFields());
    }

    @Test
    public void shippingInfoWidget_whenValidationFails_errorTextRenders() {
        mCountryAutoCompleteTextView.updateUIForCountryEntered(Locale.US.getDisplayCountry());
        mShippingInfoWidget.validateAllFields();
        assertTrue(mAddressLine1TextInputLayout.isErrorEnabled());
        assertTrue(mCityTextInputLayout.isErrorEnabled());
        assertTrue(mNameTextInputLayout.isErrorEnabled());
        assertTrue(mPostalCodeTextInputLayout.isErrorEnabled());
        assertTrue(mStateTextInputLayout.isErrorEnabled());
        mAddressLine1EditText.setText("123 Fake Address");
        mNameEditText.setText("Fake Name");
        mCityEditText.setText("Fake City");
        mPostalEditText.setText("12345");
        mStateEditText.setText("CA");
        mShippingInfoWidget.validateAllFields();
        assertFalse(mAddressLine1TextInputLayout.isErrorEnabled());
        assertFalse(mCityTextInputLayout.isErrorEnabled());
        assertFalse(mNameTextInputLayout.isErrorEnabled());
        assertFalse(mPostalCodeTextInputLayout.isErrorEnabled());
        assertFalse(mStateTextInputLayout.isErrorEnabled());
        mPostalEditText.setText("");
        mShippingInfoWidget.validateAllFields();
        assertTrue(mPostalCodeTextInputLayout.isErrorEnabled());
        mCountryAutoCompleteTextView.updateUIForCountryEntered(new Locale("", mNoPostalCodeCountry).getDisplayCountry());
        mShippingInfoWidget.validateAllFields();
        assertFalse(mStateTextInputLayout.isErrorEnabled());
    }

    @Test
    public void shippingInfoWidget_whenErrorOccurs_errorsRenderInternationalized() {
        mCountryAutoCompleteTextView.updateUIForCountryEntered(Locale.US.getDisplayCountry());
        mShippingInfoWidget.validateAllFields();
        assertEquals(mStateTextInputLayout.getError(), mShippingInfoWidget.getResources().getString(R.string.address_state_required));
        assertEquals(mPostalCodeTextInputLayout.getError(), mShippingInfoWidget.getResources().getString(R.string.address_zip_invalid));

        mCountryAutoCompleteTextView.updateUIForCountryEntered(Locale.UK.getDisplayCountry());
        mShippingInfoWidget.validateAllFields();
        assertEquals(mStateTextInputLayout.getError(), mShippingInfoWidget.getResources().getString(R.string.address_county_required));
        assertEquals(mPostalCodeTextInputLayout.getError(), mShippingInfoWidget.getResources().getString(R.string.address_postcode_invalid));

        mCountryAutoCompleteTextView.updateUIForCountryEntered( Locale.CANADA.getDisplayCountry());
        mShippingInfoWidget.validateAllFields();
        assertEquals(mStateTextInputLayout.getError(), mShippingInfoWidget.getResources().getString(R.string.address_province_required));
        assertEquals(mPostalCodeTextInputLayout.getError(), mShippingInfoWidget.getResources().getString(R.string.address_postal_code_invalid));

        mCountryAutoCompleteTextView.updateUIForCountryEntered(new Locale("", mNoPostalCodeCountry).getDisplayCountry());
        mShippingInfoWidget.validateAllFields();
        assertEquals(mStateTextInputLayout.getError(), mShippingInfoWidget.getResources().getString(R.string.address_region_generic_required));
    }

    @Test
    public void shippingInfoWidget_whenFieldsOptional_markedAsOptional(){
        assertEquals(mPostalCodeTextInputLayout.getHint().toString(), mShippingInfoWidget.getResources().getString(R.string.address_label_zip_code));
        assertEquals(mNameTextInputLayout.getHint().toString(), mShippingInfoWidget.getResources().getString(R.string.address_label_name));
        List<String> optionalFields = new ArrayList<>();
        optionalFields.add(ShippingInfoWidget.POSTAL_CODE_FIELD);
        mShippingInfoWidget.setOptionalFields(optionalFields);
        assertEquals(mPostalCodeTextInputLayout.getHint().toString(), mShippingInfoWidget.getResources().getString(R.string.address_label_zip_code_optional));
        assertEquals(mNameTextInputLayout.getHint().toString(), mShippingInfoWidget.getResources().getString(R.string.address_label_name));
        mCountryAutoCompleteTextView.updateUIForCountryEntered(Locale.CANADA.getDisplayCountry());
        assertEquals(mStateTextInputLayout.getHint().toString(), mShippingInfoWidget.getResources().getString(R.string.address_label_province));
        optionalFields.add(ShippingInfoWidget.STATE_FIELD);
        mShippingInfoWidget.setOptionalFields(optionalFields);
        assertEquals(mStateTextInputLayout.getHint().toString(), mShippingInfoWidget.getResources().getString(R.string.address_label_province_optional));
    }

    @Test
    public void shippingInfoWidget_whenFieldsHidden_renderedHidden() {
        assertEquals(mNameTextInputLayout.getVisibility(), View.VISIBLE);
        assertEquals(mPostalCodeTextInputLayout.getVisibility(), View.VISIBLE);
        List<String> hiddenFields = new ArrayList<>();
        hiddenFields.add(ShippingInfoWidget.POSTAL_CODE_FIELD);
        mShippingInfoWidget.setHiddenFields(hiddenFields);
        assertEquals(mPostalCodeTextInputLayout.getVisibility(), View.GONE);
        mCountryAutoCompleteTextView.updateUIForCountryEntered(Locale.CANADA.getDisplayCountry());
        assertEquals(mPostalCodeTextInputLayout.getVisibility(), View.GONE);
    }

    @Test
    public void getShippingInfo_whenShippingInfoInvalid_returnsNull() {
        assertNull(mShippingInfoWidget.getShippingInformation());
    }

    @Test
    public void getShippingInfo_whenShippingInfoValid_returnsExpected() {
        mStateEditText.setText("CA");
        mCityEditText.setText("San Francisco");
        mAddressLine1EditText.setText("185 Berry St");
        mAddressLine2EditText.setText("10th Floor");
        mNameEditText.setText("Fake Name");
        mPhoneEditText.setText("(123) 456 - 7890");
        mPostalEditText.setText("12345");
        mCountryAutoCompleteTextView.updateUIForCountryEntered(Locale.US.getDisplayCountry());
        ShippingInformation inputShippingInfo = mShippingInfoWidget.getShippingInformation();
        assertEquals(inputShippingInfo.toMap(), mShippingInfo.toMap());
    }

    @Test
    public void populateShippingInfo_whenShippingInfoProvided_populates() {
        mShippingInfoWidget.populateShippingInfo(mShippingInfo);
        assertEquals(mStateEditText.getText().toString(), "CA");
        assertEquals(mCityEditText.getText().toString(), "San Francisco");
        assertEquals(mAddressLine1EditText.getText().toString(), "185 Berry St");
        assertEquals(mAddressLine2EditText.getText().toString(), "10th Floor");
        assertEquals(mPhoneEditText.getText().toString(), "(123) 456 - 7890");
        assertEquals(mPostalEditText.getText().toString(), "12345");
        assertEquals(mNameEditText.getText().toString(), "Fake Name");
        assertEquals(mCountryAutoCompleteTextView.getSelectedCountryCode(), "US");
    }

}

File: stripe/src/test/java/com/stripe/android/view/ShippingInfoWidgetTest.kt
package com.stripe.android.view

import android.content.Context
import android.view.View
import androidx.test.core.app.ApplicationProvider
import com.google.android.material.textfield.TextInputLayout
import com.google.common.truth.Truth.assertThat
import com.nhaarman.mockitokotlin2.mock
import com.stripe.android.ApiKeyFixtures
import com.stripe.android.CustomerSession
import com.stripe.android.EphemeralKeyProvider
import com.stripe.android.PaymentConfiguration
import com.stripe.android.PaymentSessionData
import com.stripe.android.PaymentSessionFixtures
import com.stripe.android.R
import com.stripe.android.model.Address
import com.stripe.android.model.ShippingInformation
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import java.util.Locale
import kotlin.test.BeforeTest
import kotlin.test.Test

/**
 * Test class for [ShippingInfoWidget]
 */
@RunWith(RobolectricTestRunner::class)
class ShippingInfoWidgetTest {
    private lateinit var shippingInfoWidget: ShippingInfoWidget
    private lateinit var addressLine1TextInputLayout: TextInputLayout
    private lateinit var addressLine2TextInputLayout: TextInputLayout
    private lateinit var cityTextInputLayout: TextInputLayout
    private lateinit var nameTextInputLayout: TextInputLayout
    private lateinit var postalCodeTextInputLayout: TextInputLayout
    private lateinit var stateTextInputLayout: TextInputLayout
    private lateinit var addressLine1EditText: StripeEditText
    private lateinit var addressLine2EditText: StripeEditText
    private lateinit var postalEditText: StripeEditText
    private lateinit var cityEditText: StripeEditText
    private lateinit var nameEditText: StripeEditText
    private lateinit var stateEditText: StripeEditText
    private lateinit var phoneEditText: StripeEditText
    private lateinit var countryAutoCompleteTextView: CountryAutoCompleteTextView

    private val ephemeralKeyProvider: EphemeralKeyProvider = mock()

    private val context = ApplicationProvider.getApplicationContext<Context>()
    private val res = context.resources
    private val activityScenarioFactory = ActivityScenarioFactory(context)

    @BeforeTest
    fun setup() {
        Locale.setDefault(Locale.US)

        PaymentConfiguration.init(context, ApiKeyFixtures.FAKE_PUBLISHABLE_KEY)
        CustomerSession.initCustomerSession(context, ephemeralKeyProvider)

        val config = PaymentSessionFixtures.CONFIG.copy(
            prepopulatedShippingInfo = null,
            allowedShippingCountryCodes = emptySet()
        )
        activityScenarioFactory.create<PaymentFlowActivity>(
            PaymentFlowActivityStarter.Args(
                paymentSessionConfig = config,
                paymentSessionData = PaymentSessionData(config)
            )
        ).use { activityScenario ->
            activityScenario.onActivity {
                shippingInfoWidget = it.findViewById(R.id.shipping_info_widget)
                addressLine1TextInputLayout = shippingInfoWidget.findViewById(R.id.tl_address_line1_aaw)
                addressLine2TextInputLayout = shippingInfoWidget.findViewById(R.id.tl_address_line2_aaw)
                cityTextInputLayout = shippingInfoWidget.findViewById(R.id.tl_city_aaw)
                nameTextInputLayout = shippingInfoWidget.findViewById(R.id.tl_name_aaw)
                postalCodeTextInputLayout = shippingInfoWidget.findViewById(R.id.tl_postal_code_aaw)
                stateTextInputLayout = shippingInfoWidget.findViewById(R.id.tl_state_aaw)
                addressLine1EditText = shippingInfoWidget.findViewById(R.id.et_address_line_one_aaw)
                addressLine2EditText = shippingInfoWidget.findViewById(R.id.et_address_line_two_aaw)
                cityEditText = shippingInfoWidget.findViewById(R.id.et_city_aaw)
                nameEditText = shippingInfoWidget.findViewById(R.id.et_name_aaw)
                postalEditText = shippingInfoWidget.findViewById(R.id.et_postal_code_aaw)
                stateEditText = shippingInfoWidget.findViewById(R.id.et_state_aaw)
                phoneEditText = shippingInfoWidget.findViewById(R.id.et_phone_number_aaw)
                countryAutoCompleteTextView = shippingInfoWidget.findViewById(R.id.country_autocomplete_aaw)
            }
        }
    }

    @Test
    fun shippingInfoWidget_whenCountryChanged_fieldsRenderCorrectly() {
        countryAutoCompleteTextView.updateUiForCountryEntered(Locale.US.displayCountry)
        assertThat(addressLine1TextInputLayout.hint)
            .isEqualTo(res.getString(R.string.address_label_address))
        assertThat(addressLine2TextInputLayout.hint)
            .isEqualTo(res.getString(R.string.address_label_apt_optional))
        assertThat(postalCodeTextInputLayout.hint)
            .isEqualTo(res.getString(R.string.address_label_zip_code))
        assertThat(stateTextInputLayout.hint)
            .isEqualTo(res.getString(R.string.address_label_state))

        countryAutoCompleteTextView.updateUiForCountryEntered(Locale.CANADA.displayCountry)
        assertThat(addressLine1TextInputLayout.hint)
            .isEqualTo(res.getString(R.string.address_label_address))
        assertThat(addressLine2TextInputLayout.hint)
            .isEqualTo(res.getString(R.string.address_label_apt_optional))
        assertThat(postalCodeTextInputLayout.hint)
            .isEqualTo(res.getString(R.string.address_label_postal_code))
        assertThat(stateTextInputLayout.hint)
            .isEqualTo(res.getString(R.string.address_label_province))

        countryAutoCompleteTextView.updateUiForCountryEntered(Locale.UK.displayCountry)
        assertThat(addressLine1TextInputLayout.hint)
            .isEqualTo(res.getString(R.string.address_label_address_line1))
        assertThat(addressLine2TextInputLayout.hint)
            .isEqualTo(res.getString(R.string.address_label_address_line2_optional))
        assertThat(postalCodeTextInputLayout.hint)
            .isEqualTo(res.getString(R.string.address_label_postcode))
        assertThat(stateTextInputLayout.hint)
            .isEqualTo(res.getString(R.string.address_label_county))

        countryAutoCompleteTextView.updateUiForCountryEntered(
            Locale("", NO_POSTAL_CODE_COUNTRY_CODE).displayCountry
        )
        assertThat(addressLine1TextInputLayout.hint)
            .isEqualTo(res.getString(R.string.address_label_address_line1))
        assertThat(addressLine2TextInputLayout.hint)
            .isEqualTo(res.getString(R.string.address_label_address_line2_optional))
        assertThat(postalCodeTextInputLayout.visibility)
            .isEqualTo(View.GONE)
        assertThat(stateTextInputLayout.hint)
            .isEqualTo(res.getString(R.string.address_label_region_generic))
    }

    @Test
    fun shippingInfoWidget_addressSaved_validationTriggers() {
        countryAutoCompleteTextView.updateUiForCountryEntered(Locale.US.displayCountry)
        assertThat(shippingInfoWidget.validateAllFields())
            .isFalse()
        addressLine1EditText.setText("123 Fake Address")
        nameEditText.setText("Fake Name")
        cityEditText.setText("Fake City")
        postalEditText.setText("12345")
        stateEditText.setText("CA")
        phoneEditText.setText("(123) 456 - 7890")
        assertThat(shippingInfoWidget.validateAllFields())
            .isTrue()
        postalEditText.setText("")
        assertThat(shippingInfoWidget.validateAllFields())
            .isFalse()
        postalEditText.setText("ABCDEF")
        assertThat(shippingInfoWidget.validateAllFields())
            .isFalse()
        countryAutoCompleteTextView
            .updateUiForCountryEntered(Locale("", NO_POSTAL_CODE_COUNTRY_CODE).displayCountry)
        assertThat(shippingInfoWidget.validateAllFields())
            .isTrue()
    }

    @Test
    fun shippingInfoWidget_whenValidationFails_errorTextRenders() {
        countryAutoCompleteTextView.updateUiForCountryEntered(Locale.US.displayCountry)
        assertThat(shippingInfoWidget.validateAllFields())
            .isFalse()

        assertThat(addressLine1TextInputLayout.isErrorEnabled)
            .isTrue()
        assertThat(cityTextInputLayout.isErrorEnabled)
            .isTrue()
        assertThat(nameTextInputLayout.isErrorEnabled)
            .isTrue()
        assertThat(postalCodeTextInputLayout.isErrorEnabled)
            .isTrue()
        assertThat(stateTextInputLayout.isErrorEnabled)
            .isTrue()
        addressLine1EditText.setText("123 Fake Address")
        nameEditText.setText("Fake Name")
        cityEditText.setText("Fake City")
        postalEditText.setText("12345")
        stateEditText.setText("CA")

        assertThat(shippingInfoWidget.validateAllFields())
            .isTrue()

        assertThat(addressLine1TextInputLayout.isErrorEnabled)
            .isFalse()
        assertThat(cityTextInputLayout.isErrorEnabled)
            .isFalse()
        assertThat(nameTextInputLayout.isErrorEnabled)
            .isFalse()
        assertThat(postalCodeTextInputLayout.isErrorEnabled)
            .isFalse()
        assertThat(stateTextInputLayout.isErrorEnabled)
            .isFalse()
        postalEditText.setText("")

        assertThat(shippingInfoWidget.validateAllFields())
            .isFalse()
        assertThat(postalCodeTextInputLayout.isErrorEnabled)
            .isTrue()

        countryAutoCompleteTextView
            .updateUiForCountryEntered(
                Locale("", NO_POSTAL_CODE_COUNTRY_CODE).displayCountry
            )
        assertThat(shippingInfoWidget.validateAllFields())
            .isTrue()
        assertThat(stateTextInputLayout.isErrorEnabled)
            .isFalse()
    }

    @Test
    fun shippingInfoWidget_whenErrorOccurs_errorsRenderInternationalized() {
        countryAutoCompleteTextView.updateUiForCountryEntered(Locale.US.displayCountry)
        assertThat(shippingInfoWidget.validateAllFields())
            .isFalse()

        assertThat(stateTextInputLayout.error)
            .isEqualTo(res.getString(R.string.address_state_required))
        assertThat(postalCodeTextInputLayout.error)
            .isEqualTo(res.getString(R.string.address_zip_invalid))

        countryAutoCompleteTextView.updateUiForCountryEntered(Locale.UK.displayCountry)
        assertThat(shippingInfoWidget.validateAllFields())
            .isFalse()
        assertThat(stateTextInputLayout.error)
            .isEqualTo(res.getString(R.string.address_county_required))
        assertThat(postalCodeTextInputLayout.error)
            .isEqualTo(res.getString(R.string.address_postcode_invalid))

        countryAutoCompleteTextView.updateUiForCountryEntered(Locale.CANADA.displayCountry)
        assertThat(shippingInfoWidget.validateAllFields())
            .isFalse()
        assertThat(stateTextInputLayout.error)
            .isEqualTo(res.getString(R.string.address_province_required))
        assertThat(postalCodeTextInputLayout.error)
            .isEqualTo(res.getString(R.string.address_postal_code_invalid))

        countryAutoCompleteTextView
            .updateUiForCountryEntered(Locale("", NO_POSTAL_CODE_COUNTRY_CODE).displayCountry)
        assertThat(shippingInfoWidget.validateAllFields())
            .isFalse()
        assertThat(stateTextInputLayout.error)
            .isEqualTo(res.getString(R.string.address_region_generic_required))
    }

    @Test
    fun shippingInfoWidget_whenFieldsOptional_markedAsOptional() {
        assertThat(postalCodeTextInputLayout.hint.toString())
            .isEqualTo(res.getString(R.string.address_label_zip_code))
        assertThat(nameTextInputLayout.hint.toString())
            .isEqualTo(res.getString(R.string.address_label_name))
        shippingInfoWidget.optionalFields = listOf(
            ShippingInfoWidget.CustomizableShippingField.PostalCode
        )
        assertThat(postalCodeTextInputLayout.hint.toString())
            .isEqualTo(res.getString(R.string.address_label_zip_code_optional))
        assertThat(nameTextInputLayout.hint.toString())
            .isEqualTo(res.getString(R.string.address_label_name))
        countryAutoCompleteTextView.updateUiForCountryEntered(Locale.CANADA.displayCountry)
        assertThat(stateTextInputLayout.hint.toString())
            .isEqualTo(res.getString(R.string.address_label_province))
        shippingInfoWidget.optionalFields = listOf(
            ShippingInfoWidget.CustomizableShippingField.PostalCode,
            ShippingInfoWidget.CustomizableShippingField.State
        )
        assertThat(stateTextInputLayout.hint.toString())
            .isEqualTo(res.getString(R.string.address_label_province_optional))
    }

    @Test
    fun shippingInfoWidget_whenFieldsHidden_renderedHidden() {
        assertThat(nameTextInputLayout.visibility)
            .isEqualTo(View.VISIBLE)
        assertThat(postalCodeTextInputLayout.visibility)
            .isEqualTo(View.VISIBLE)
        shippingInfoWidget.hiddenFields = listOf(
            ShippingInfoWidget.CustomizableShippingField.PostalCode
        )
        assertThat(postalCodeTextInputLayout.visibility)
            .isEqualTo(View.GONE)
        countryAutoCompleteTextView.updateUiForCountryEntered(
            Locale.CANADA.displayCountry
        )
        assertThat(postalCodeTextInputLayout.visibility)
            .isEqualTo(View.GONE)
    }

    @Test
    fun getShippingInfo_whenShippingInfoInvalid_returnsNull() {
        assertThat(shippingInfoWidget.shippingInformation)
            .isNull()
    }

    @Test
    fun getShippingInfo_whenShippingInfoValid_returnsExpected() {
        stateEditText.setText("CA")
        cityEditText.setText("San Francisco")
        addressLine1EditText.setText("185 Berry St")
        addressLine2EditText.setText("10th Floor")
        nameEditText.setText("Fake Name")
        phoneEditText.setText("(123) 456 - 7890")
        postalEditText.setText("12345")
        countryAutoCompleteTextView.updateUiForCountryEntered(Locale.US.displayCountry)

        assertThat(shippingInfoWidget.shippingInformation)
            .isEqualTo(SHIPPING_INFO)
    }

    @Test
    fun populateShippingInfo_whenShippingInfoProvided_populates() {
        shippingInfoWidget.populateShippingInfo(SHIPPING_INFO)
        assertThat(stateEditText.fieldText)
            .isEqualTo("CA")
        assertThat(cityEditText.fieldText)
            .isEqualTo("San Francisco")
        assertThat(addressLine1EditText.fieldText)
            .isEqualTo("185 Berry St")
        assertThat(addressLine2EditText.fieldText)
            .isEqualTo("10th Floor")
        assertThat(phoneEditText.fieldText)
            .isEqualTo("(123) 456 - 7890")
        assertThat(postalEditText.fieldText)
            .isEqualTo("12345")
        assertThat(nameEditText.fieldText)
            .isEqualTo("Fake Name")
        assertThat(countryAutoCompleteTextView.selectedCountry?.code)
            .isEqualTo("US")
    }

    @Test
    fun getSelectedShippingCountry_whenShippingInfoProvided_returnsExpected() {
        shippingInfoWidget.setAllowedCountryCodes(setOf("US", "CA"))
        shippingInfoWidget.populateShippingInfo(SHIPPING_INFO_CA)
        assertThat(stateEditText.fieldText)
            .isEqualTo("Ontario")
        assertThat(cityEditText.fieldText)
            .isEqualTo("Ontario")
        assertThat(addressLine1EditText.fieldText)
            .isEqualTo("185 Berry St")
        assertThat(addressLine2EditText.fieldText)
            .isEqualTo("10th Floor")
        assertThat(phoneEditText.fieldText)
            .isEqualTo("416-759-0260")
        assertThat(postalEditText.fieldText)
            .isEqualTo("M4B1B5")
        assertThat(nameEditText.fieldText)
            .isEqualTo("Fake Name")
        assertThat(countryAutoCompleteTextView.selectedCountry?.code)
            .isEqualTo("CA")
    }

    private companion object {
        private const val NO_POSTAL_CODE_COUNTRY_CODE = "ZW" // Zimbabwe

        private val SHIPPING_INFO = ShippingInformation(
            Address.Builder()
                .setCity("San Francisco")
                .setState("CA")
                .setCountry("US")
                .setLine1("185 Berry St")
                .setLine2("10th Floor")
                .setPostalCode("12345")
                .build(),
            "Fake Name",
            "(123) 456 - 7890"
        )

        private val SHIPPING_INFO_CA = ShippingInformation(
            Address.Builder()
                .setCity("Ontario")
                .setState("Ontario")
                .setCountry("CA")
                .setLine1("185 Berry St")
                .setLine2("10th Floor")
                .setPostalCode("M4B1B5")
                .build(),
            "Fake Name",
            "416-759-0260"
        )
    }
}
